### ISSUE TITLE
Static RPC's that are on a generic class don't work

### ISSUE THREAD

**Issue by Remscar** [Facepunch/sbox-issues#6629]
### Describe the bug

I have a static RPC on a class, which happens to be a generic class.

When I call the RPC I get the following error on a client;

```
[Generic] Late bound operations cannot be performed on types or methods for which ContainsGenericParameters is true.

System.InvalidOperationException: Late bound operations cannot be performed on types or methods for which ContainsGenericParameters is true.
   at System.Reflection.RuntimeMethodInfo.ThrowNoInvokeException()
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Sandbox.MethodDescription.Invoke(Object targetObject, Object[] parameters) in C:\build\_work\sbox\sbox\engine\Sandbox.Reflection\TypeLibrary\Members\MethodDescription.cs:line 93
   at Sandbox.Rpc.HandleIncoming(StaticRpcMsg message, Connection source) in C:\build\_work\sbox\sbox\engine\Sandbox.Game\Scene\Networking\Rpc.cs:line 310
   at Sandbox.SceneNetworkSystem.OnStaticRpc(StaticRpcMsg message, Connection source, Guid msgid) in C:\build\_work\sbox\sbox\engine\Sandbox.Game\Scene\Networking\SceneNetworkSystem.cs:line 402
   at Sandbox.Network.NetworkSystem.<>c__DisplayClass82_0`1.<AddHandler>b__0(Object o, Connection channel, Guid g) in C:\build\_work\sbox\sbox\engine\Sandbox.Engine\Networking\System\NetworkSystem.Message.cs:line 36
   at Sandbox.Network.NetworkSystem.HandleIncomingMessage(NetworkMessage msg) in C:\build\_work\sbox\sbox\engine\Sandbox.Engine\Networking\System\NetworkSystem.Message.cs:line 245
```

What's interesting is calling the RPC on the host (to the host) works fine.

### To Reproduce

Use this project, launch a new instance, press mouse 1 on the host, look at the console on the attached instance.

https://github.com/Remscar/sandbox-tester/tree/bug/rpc-1

Alternatively use this code.

```cs
using System;
using System.Threading.Tasks;

public sealed class ClickToSend : Component
{

	protected override void OnFixedUpdate()
	{
		if (IsProxy) return;

		if (Input.Pressed("attack1"))
		{
			StaticWithRPCs<int>.DoSendMessage();
		}
	}

}


public class StaticWithRPCs<T> where T : unmanaged
{

	public static void DoSendMessage()
	{
		using (Rpc.FilterExclude(Connection.Host))
		{
			SendMessage(Guid.NewGuid());
		}
	}

	[Broadcast]
	public static void SendMessage(Guid someGuid)
	{
		Log.Info($"Got message: {someGuid}");
	}
}
```

### Expected behavior

Either the RPC should work, or it should give a better error saying why it might not be working.

What's really strange to me is I swear I had this working fine a few days ago, but suddenly it stopped working.

The error message was confusing and took me nearly 30 minutes of pulling out my hair (I have none left) to finally figure out how to reproduce it, I doubt I will be the only person who runs into this.

**Comment by DevulTj** [comment]
I don't get why you want to use this, this seems really unnecessary.

**Comment by Remscar** [comment]
> I don't get why you want to use this, this seems really unnecessary.

@DevulTj I'm happy to explain why I would want to use this;

In my use case I have a generic class which I am utilizing the [curiously recurring template pattern](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern).

For instance I have an Upgrade<T> class where I derive different types of upgrades;

```cs
public abstract class Upgrade<T> : IUpgrade where T : Upgrade<T>, IUpgrade, new() {
  // some code

  public static T Create( Team team ) {
      return new T() {
        Team = team,
      };
    }

  public virtual T WithName( string name ) {
    Name = name;
    return (T)this;
  }
}

class MagicUpgrade : Upgrade<MagicUpgrade>{
  // some magic upgrade specific code

  public MagicUpgrade BoostsManaPerLevel(int amount) {
    ManaBoostPerLevel = amount;
    return this;
  }
}
```

The reason I do this is because I want to use a Builder pattern to construct instances of upgrades

```cs
var upgrade = MagicUpgrade.Create().WithName("foo").BoostsManaPerLevel(10); // and so on
```

And since I have many different types of upgrades, which may derive from that base upgrade, I want to be able to define  builder methods on that base generic class

```cs
public virtual T WithName( string name ) {
  Name = name;
  return (T)this;
}
```

To make this chain together like in the above sample I have to use the CRTP, which means generic classes. Otherwise I have to cast it to the derived type after I call a base type method.

Eventually as I began adding networking code, I wanted to send RPC's to tell clients about some event relating to the upgrade (for instance the upgrade level increased.) All of my logic is already in these generic classes, and so I made a static method for the RPC. However that's where I ran into this issue where static RPC's don't work.

I can understand if this is too much work to implement support for, but it would be nice at the very least to have some mention of this limitation in the docs, or a better error message when this is encountered.

My workaround is to have a non-generic static class with my RPC methods, which is working fine, but is not quite the ideal since:
- I'm unable to make the RPC methods private only to `Upgrade<T>` derived classes
- I'm unable to call private/protected methods inside of `Upgrade<T>`
- The relevant code is split between two areas
